kernel/univ.ml

```diff
@@ -755,6 +755,8 @@ struct
     | Invariant, _ | _, Invariant -> Invariant
     | Covariant, Covariant -> Covariant
 
+  let equal : t -> t -> bool = (=)
```

Comment:
I know this looks obsessive, but please write this as a pattern matching instead of using the pervasive equality.

Reply:
Isn't there a way to derive these equalities? I'm going to need an equality on vernac ASTs soon, it is going to be very tedious to write by hand.

Reply:
There are PPXs for that, but that would require adding dependencies.

Reply:
>I know this looks obsessive, but please write this as a pattern matching instead of using the pervasive equality.

Why though?

Reply:
> that would require adding dependencies

Is that a problem? I now this is a bit beyond the scope of this PR, but there are hundreds of similar examples in the codebase.

Reply:
@SkySkimmer using pervasive equality has quite a few downsides:
- it is not efficient (shouldn't matter here though)
- it is not future proof, e.g. if you start adding constructors that contain non-structural-equality-using data, and the compiler won't help
- it prevents easily checking that you're not relying on pervasive equality for opaque data

Reply:
@maximedenes I'd personally favour staying out of the PPX hell as long as we can. This mechanism is still fragile, and while writing an equality function is tedious, it's something that needs to be done once and for all, and this will still compile on the OCaml versions from the year 2100.

(Also, how are you going to handle dynamic extensions to the Vernac AST?)

Reply:
> it's something that needs to be done once and for all

I thought your concerns above were precisely maintenance. If you add a constructor, you also need to update your equality function. No need to say you can get it wrong.

Having no meta-programming facilities feels a bit like we are lagging 50 years behind. If you compare it to programming in Rust...

> (Also, how are you going to handle dynamic extensions to the Vernac AST?)

As usual, via a dedicated node and a registration mechanism.

Reply:
> Having no meta-programming facilities feels a bit like we are lagging 50 years behind. If you compare it to programming in Rust...

The main difference being that the derive functions are part of the official Rust language. I'd suggest loudly complaining to the OCaml devs...